catroid/src/main/java/org/catrobat/catroid/formulaeditor/datacontainer/SpriteDataBehaviour.java

```diff
@@ -0,0 +1,185 @@
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.formulaeditor.datacontainer;
+
+import org.catrobat.catroid.content.Scene;
+import org.catrobat.catroid.content.Sprite;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+
+abstract class SpriteDataBehaviour<K, V> {
+
+	protected abstract void reset(List<V> dataList);
+	protected abstract Map<K, List<V>> getDataMap();
+	protected abstract V newInstance(String name);
+	protected abstract V newInstance(String name, Object value);
+	protected abstract String getDataName(V data);
+	protected abstract void setDataName(V data, String name);
+	protected abstract String getKeyName(K key);
+	protected abstract void setValue(V dataToAdd, Object value);
+	protected abstract boolean isClone(K key);
+
+	V add(K key, String name) {
+		List<V> dataList = getOrCreate(key);
+		V dataToAdd = newInstance(name);
+		dataList.add(dataToAdd);
+		return dataToAdd;
+	}
+
+	V add(K key, String name, Object value) {
+		List<V> dataList = getOrCreate(key);
+		V dataToAdd = newInstance(name, value);
+		dataList.add(dataToAdd);
+		return dataToAdd;
+	}
+
+	V addIfNotExists(K key, String name) {
+		List<V> list = getOrCreate(key);
+
+		V data = find(key, name);
+		if (data != null) {
+			return data;
+		}
+
+		V addedData = newInstance(name);
+		list.add(addedData);
+		getDataMap().put(key, list);
+
+		return addedData;
+	}
+
+	V addIfNotExists(K key, String name, Object value) {
+		V dataToAdd = addIfNotExists(key, name);
+		setValue(dataToAdd, value);
+		return dataToAdd;
+	}
+
+	V get(K key, String name) {
+		return find(key, name);
+	}
+
+	List<V> getOrCreate(K key) {
+		List<V> data = getDataMap().get(key);
+
+		if (data == null) {
+			data = new ArrayList<>();
+			getDataMap().put(key, data);
+		}
+		return data;
+	}
+
+	V find(K key, String name) {
+		List<V> dataList = getOrCreate(key);
+
+		for (V data : dataList) {
+			if (getDataName(data).equals(name)) {
+				return data;
+			}
+		}
+		return null;
+	}
+
+	boolean exists(K key, V value) {
+		List<V> list = getDataMap().get(key);
+		return list != null && list.contains(value);
+	}
+
+	boolean exists(K key, String name) {
+		if (key == null) {
```

Comment:
However, I think it shouldn't normally happen as this is only called within a sprite (rename and create dialogs). Otherwise it'll be another bug - so I'll remove that check ;)